### PR TITLE
Implement window list for wf-panel

### PR DESCRIPTION
**Pull request by soreau**
This is a basic window list implementation for wf-panel using the foreign toplevel protocol. It features icons, title, a right click menu to control windows and dnd to reorder buttons in the list.

### CODE REVIEW

File: src/panel/meson.build [hunk 1]
```diff
@@ -4,7 +4,9 @@ widget_sources = ['widgets/battery.cpp',
                   'widgets/launchers.cpp',
                   'widgets/network.cpp',
                   'widgets/spacing.cpp',
-                  'widgets/wf-popover.cpp']
+                  'widgets/wf-popover.cpp',
+                  'widgets/window-list.cpp',
+                  'widgets/toplevel.cpp']
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/meson.build]: I put the relevant files in a window-list subdirectory.

File: src/panel/widgets/toplevel.cpp [hunk 2]
```diff
@@ -0,0 +1,603 @@
+#include <gdkmm/seat.h>
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/toplevel.cpp]: Done.

File: src/panel/widgets/toplevel.cpp [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+#include <gtkmm/gesturedrag.h>
+#include <giomm/desktopappinfo.h>
+#include <iostream>
+
+#include "toplevel.hpp"
+#include "gtk-utils.hpp"
+#include "panel.hpp"
+#include <cassert>
+
+#define DND_BIT_FORMAT 8
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/toplevel.cpp]: Fixed.

File: src/panel/widgets/toplevel.cpp [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+class WayfireToplevel::impl
+{
+    zwlr_foreign_toplevel_handle_v1 *handle;
+    uint32_t state;
+
+    Gtk::Button button;
+    Gtk::HBox button_contents;
+    Gtk::Image image;
+    Gtk::Label label;
+    Gtk::Menu *menu;
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/toplevel.cpp]: Done.

File: src/panel/widgets/toplevel.cpp [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+    {
+        int x = _x + grab_start_x;
+        int y = _y + grab_start_y;
+        int width = button.get_allocated_width();
+        int height = button.get_allocated_height();
+
+        window_list->box.set_top_widget(nullptr);
+        set_flat_class(!(state & WF_TOPLEVEL_STATE_ACTIVATED));
+
+        if (x < 0 || x > width || y < 0 || y > height)
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/toplevel.cpp]: Done.

File: src/panel/widgets/toplevel.cpp [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+            sigc::mem_fun(this, &WayfireToplevel::impl::on_allocation_changed));
+        button.property_scale_factor().signal_changed()
+            .connect(sigc::mem_fun(this, &WayfireToplevel::impl::on_scale_update));
+        button.signal_button_press_event().connect(
+            sigc::mem_fun(this, &WayfireToplevel::impl::on_button_press_event));
+
+        minimize.set_label("Minimize");
+        maximize.set_label("Maximize");
+        close.set_label("Close");
+        minimize.signal_button_press_event().connect(
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/toplevel.cpp]: Done.

File: src/panel/widgets/toplevel.cpp [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+    void handle_output_leave(wl_output *output)
+    {
+        auto container = &window_list->box;
+        if (window_list->output->handle == output)
+            container->remove(button);
+    }
+
+    void handle_toplevel_closed()
+    {
+        delete menu;
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/toplevel.cpp]: Done.

File: src/panel/widgets/window-list.cpp [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+    wl_registry *registry = wl_display_get_registry(display);
+    wl_registry_add_listener(registry, &registry_listener, this);
+    wl_display_roundtrip(display);
+
+    if (!this->manager)
+    {
+        std::cerr << "Compositor doesn't support" <<
+            " wlr-foreign-toplevel-management, exiting." << std::endl;
+        std::exit(-1);
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/window-list.cpp]: Done.

File: src/panel/widgets/window-list.cpp [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+    return panel_width * DEFAULT_SIZE_PC;
+}
+
+int WayfireWindowList::get_target_button_width()
+{
+    int num_children = box.get_children().size();
+    int target_width = get_default_button_width();
+
+    return target_width;
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/window-list.cpp]: Done.

File: src/panel/widgets/toplevel.cpp [hunk 10]
```diff
@@ -0,0 +1,603 @@
...
+            container->add(button);
+            container->show_all();
+        }
+
+        update_menu_item_text();
+    }
+
+    void handle_output_leave(wl_output *output)
+    {
+        auto container = &window_list->box;
```
[the remark this answers is not in the document]
Comment (soreau) [reply src/panel/widgets/toplevel.cpp]: Done.

**Comment by ammen99** [comment]
@soreau Thanks for the work on this!

**Pull request merged**